### PR TITLE
Do not run functions with necessary side-effects in assert()

### PR DESCRIPTION
**Pull request by gmaxwell**

### CODE REVIEW

**Review by dcousens — APPROVED** [review]

**Comment by dcousens** [comment]
Does the compiler,  when removing asserts,  remove the side effects too?

**Comment by sipa** [comment]
> Does the compiler, when removing asserts, remove the side effects too?

Not the compiler, but assert(X); is defined as nothing in the C library headers when compiling without debug.

**Comment by dcousens** [comment]
> s/compiler/C library headers

Right,  ouch.

**Comment by luke-jr** [comment]
Is this going to produce "variable set but not used" warnings when compiled with NDEBUG? Is there a clean way to avoid those?

**Comment by gmaxwell** [comment]
Yes, turn off that nearly useless warning. :P   The way to avoid it in gcc is to cast the variable to void in a statement of its own-- but doing so makes MSVC warn. There is no winning with overly paternalistic compiler warnings.  Besides, being silently incorrect is a lot worse than a warning.

**Comment by paveljanik** [comment]
@luke-jr Is it producing this warning? In that case, the macro `assert` is written wrong.

But, anyway: we do not support compilation with `NDEBUG` ;-)

**Comment by paveljanik** [comment]
ACK da9cdd2

**Comment by laanwj** [comment]
utACK https://github.com/bitcoin/bitcoin/commit/da9cdd2c9cb58e3737b727c21f070f074b1247a7

This is not the first time side-effects snuck into assertions, and were subsequently removed.
How can we avoid these from being introduced again?

File: src/net_processing.cpp [hunk 1]
```diff
@@ -1518,7 +1518,8 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
         }
 
         CBlock block;
-        assert(ReadBlockFromDisk(block, it->second, chainparams.GetConsensus()));
+        bool ret = ReadBlockFromDisk(block, it->second, chainparams.GetConsensus());
```
Comment (laanwj) [inline src/net_processing.cpp]: Was introduced in https://github.com/bitcoin/bitcoin/commit/da9cdd2c9cb58e3737b727c21f070f074b1247a7 by @TheBlueMatt

File: src/net_processing.cpp [hunk 2]
```diff
@@ -2730,7 +2731,8 @@ bool SendMessages(CNode* pto, CConnman& connman)
                             vHeaders.front().GetHash().ToString(), pto->id);
                     //TODO: Shouldn't need to reload block from disk, but requires refactor
                     CBlock block;
-                    assert(ReadBlockFromDisk(block, pBestIndex, consensusParams));
```
Comment (laanwj) [inline src/net_processing.cpp]: Introduced in same commit as above.

File: src/validation.cpp [hunk 3]
```diff
@@ -2100,7 +2100,8 @@ bool static DisconnectTip(CValidationState& state, const CChainParams& chainpara
         CCoinsViewCache view(pcoinsTip);
         if (!DisconnectBlock(block, state, pindexDelete, view))
             return error("DisconnectTip(): DisconnectBlock %s failed", pindexDelete->GetBlockHash().ToString());
-        assert(view.Flush());
```
Comment (laanwj) [inline src/validation.cpp]: Introduced by @sipa in 75f51f2a63e0ebe34ab290c2b7141dd240b98c3b

File: src/validation.cpp [hunk 4]
```diff
@@ -2189,7 +2190,8 @@ bool static ConnectTip(CValidationState& state, const CChainParams& chainparams,
         }
         nTime3 = GetTimeMicros(); nTimeConnectTotal += nTime3 - nTime2;
         LogPrint("bench", "  - Connect total: %.2fms [%.2fs]\n", (nTime3 - nTime2) * 0.001, nTimeConnectTotal * 0.000001);
-        assert(view.Flush());
+        bool flushed = view.Flush();
+        assert(flushed);
```
Comment (laanwj) [inline src/validation.cpp]: Introduced in same commit as above.

**Review by laanwj — APPROVED** [review]

**Pull request merged**